master/src/main/scala/org/apache/celeborn/service/deploy/master/Master.scala

```diff
@@ -678,16 +679,9 @@ private[celeborn] class Master(
         if (requestSlots.maxWorkers <= 0) slotsAssignMaxWorkers
         else Math.min(slotsAssignMaxWorkers, requestSlots.maxWorkers)),
       numAvailableWorkers)
-    val startIndex = Random.nextInt(numAvailableWorkers)
-    val selectedWorkers = new util.ArrayList[WorkerInfo](numWorkers)
-    selectedWorkers.addAll(availableWorkers.subList(
-      startIndex,
-      Math.min(numAvailableWorkers, startIndex + numWorkers)))
-    if (startIndex + numWorkers > numAvailableWorkers) {
-      selectedWorkers.addAll(availableWorkers.subList(
-        0,
-        startIndex + numWorkers - numAvailableWorkers))
-    }
+    Collections.rotate(availableWorkers, Random.nextInt(numAvailableWorkers))
```

Comment:
Thanks, LGTM for this selection way. 
It would be better if we could judge whether numAvailableWorkers and numWorkers are equal to decide whether to do the select. 
Because if there is no intentional limit on slotsAssignMaxWorkers(by default 10000) to be smaller than the number of cluster workers, the values are likely to be equal.

Reply:
@onebox-li, make sense
```
// select workers from availableWorkers randomly
Collections.rotate(availableWorkers, Random.nextInt(numAvailableWorkers))
val selectedWorkers =
  if (numWorkers == numAvailableWorkers) availableWorkers
  else availableWorkers.subList(0, numWorkers)
```

Reply:
Maybe also no need to do the rotate?

Reply:
> Maybe also no need to do the rotate?

Yeah, if remove the rotate, the slots offering could not be balanced in all workers.

Reply:
As @waitinfuture mentioned in https://github.com/apache/incubator-celeborn/pull/1975, Master just randomly picks a range of available workers.

Reply:
When allocate, it will start from a random index of worker. IMO it's fine. Refer to https://github.com/apache/incubator-celeborn/blob/main/master/src/main/java/org/apache/celeborn/service/deploy/master/SlotsAllocator.java#L339

Reply:
> When allocate, it will start from a random index of worker. IMO it's fine. Refer to https://github.com/apache/incubator-celeborn/blob/main/master/src/main/java/org/apache/celeborn/service/deploy/master/SlotsAllocator.java#L339

Right, the SlotsAllocator use a random index of workers, it start from selected workers).

If the client set the num of `requestSlots.maxWorkers` less then the num of all wokers, the master will always choose the top workers.

Reply:
> Oh I means we could also skip `rotate` in case which numWorkers == numAvailableWorkers is true.

Emm, I think it's better to keep the same logic as before, because we can not make sure the logic in `SlotsAllocator` is always offer slots randomly.